solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java

```diff
@@ -146,68 +155,140 @@ public SolrQueryRequest getRequest() {
   @Override
   public ModelWeight createWeight(IndexSearcher searcher, boolean needsScores, float boost)
       throws IOException {
-    final FeatureWeight[] allFeatureWeights = createWeights(meta.getAllFeatures(),
-        searcher, needsScores);
-    final FeatureWeight[] modelFeaturesWeights = createWeights(meta.getFeatures(),
-        searcher, needsScores);
 
-    return new ModelWeight(searcher, modelFeaturesWeights, allFeatureWeights);
-  }
+    final Collection<Feature> modelFeatures = meta.getFeatures();
+    final Collection<Feature> allFeatures = meta.getAllFeatures();
+
+    int modelFeatSize = modelFeatures.size();
+    int allFeatSize = this.extractAllFeatures ? allFeatures.size() : modelFeatSize;
 
-  private FeatureWeight[] createWeights(Collection<Feature> features,
-      IndexSearcher searcher, boolean needsScores) throws IOException {
-    final FeatureWeight[] arr = new FeatureWeight[features.size()];
-    int i = 0;
+    final FeatureWeight[] extractedFeatureWeights = new FeatureWeight[allFeatSize];
+    final FeatureWeight[] modelFeaturesWeights = new FeatureWeight[modelFeatSize];
+    createWeightsSelectively(allFeatures,modelFeatures,searcher, needsScores, extractedFeatureWeights,modelFeaturesWeights); 
+
+    int modelFeatIndex = 0;
+    return new ModelWeight(searcher, modelFeaturesWeights, extractedFeatureWeights, allFeatures.size());
+  }
+  
+  private void createWeightsSelectively(Collection<Feature> allFeatures, 
+      Collection<Feature> modelFeatures,
+      IndexSearcher searcher, boolean needsScores, FeatureWeight[] allFeatureWeights,
+      FeatureWeight[] modelFeaturesWeights) throws IOException {
+    int i = 0, j = 0;
     final SolrQueryRequest req = getRequest();
     // since the feature store is a linkedhashmap order is preserved
-    for (final Feature f : features) {
-      try {
-        final FeatureWeight fw = f.createWeight(searcher, needsScores, req, originalQuery, efi);
-
-        arr[i] = fw;
-        ++i;
-      } catch (final Exception e) {
-        throw new FeatureException("Exception from createWeight for " + f.toString() + " "
-            + e.getMessage(), e);
+    if (this.extractAllFeatures) {
+      boolean[] modelFeatMap = new boolean[allFeatures.size()];
```

Comment:
I find it odd that you pass in the model/all FeatureWeights into this function, which  have been pre-sized by the caller, but then your code here has a direct dependence on the size chosen by extractAllFeatures vs using what's directly passed in.  It makes this helper function brittle.
I think it might make more sense to just merge the sizing logic and this function into 1 so you can reuse the same size variables to keep everything a little more sane.